packages/@aws-cdk/cdk/lib/aspects/aspect.ts

```diff
@@ -0,0 +1,54 @@
+import { IConstruct } from '../core/construct';
+
+/**
+ * Represents an Aspect
+ */
+export interface IAspect {
+  /**
+   * The type of Aspect
+   */
+  readonly type: string;
+
+  /**
+   * All aspects can visit by IConstructs
+   */
+  visit(node: IConstruct): void;
+}
+
+/**
+ * The representation of an Apect
+ */
+export abstract class Aspect implements IAspect {
+  /**
+   * The type of the aspect
+   */
+  public abstract readonly type: string;
+
+  private readonly visitedBy: {[id: string]: boolean} = {};
+
+  /**
+   * The visit function is invoked during synthesis for each aspect
+   *
+   * The visit function will visit each child node in the construct tree. Each
+   * Node will only be visited once.
+   */
+  public visit(construct: IConstruct): void {
+    if (this.visitedBy[construct.node.uniqueId] === true) {
```

Comment:
I’d rename this to “visitTree” and “visitAction” to “visit”

Reply:
A little clarification. Does this also mean change:

```ts
/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * The type of Aspect
   */
  readonly type: string;

  /**
   * All aspects can visit by IConstructs
   */
  visit(node: IConstruct): void;
}
```

to 

```ts
/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * The type of Aspect
   */
  readonly type: string;

  /**
   * All aspects can visit the construct tree starting from this construct
   */
  visitTree(node: IConstruct): void;
}
```

Reply:
I wouldn’t say so. I think aspects are only expected to deal with their own visit. The tree visitation is handled by the framework, no?